blog/_posts/06-03-blogpost.md

```diff
@@ -0,0 +1,164 @@
+
+This means that if we have more than 64 bits, we index into the pointer, and if we don't, we simply deference it.  So we have a lot of code like:
+```c++
+if (num_bits > SHORT_THRESHOLD) {
+	return (BitSetPtr()[field_id] & (static_cast<field_t>(1) << pos_id));
+}
+return (*BitSetPtr().Raw() & (static_cast<field_t>(1) << pos_id));
+```
+which is the Get function.  
+
+So what could go wrong with this?  Well, the fact that we are messing with pointers means that there will inevitably be some memory errors.  Memory leaks, accidental dereferences, that sort of thing.  And so there were, mainly with earlier versions of BitSetPtr and the move and copy constructors.  Resize also posed a bit of an issue, since it is possible to cross 64 bits and go from long to short or from short to long, which requires some adjustment of how things are stored.  
+
+Once I had gotten this done, the speedups were generally quite high.
+
+![WIP BitVector Speedups All](BitVectorSpeedupsAll.png "BitVector Speedups All")
+
+![WIP BitVector Speedups](BitVectorSpeedups.png "BitVector Speedups")
+
+\<\<\<Section where I talk about the results in more detail\>\>\>
+
+## The DataMap Reference Capability
+This is how a DataMap used to be structured:
+
+![WIP DataMap Image](DataMap.png "DataMap")
+
+And this is how I have restructured it.
+
+![WIP New DataMap Image](NewDataMap.png "New DataMap")
+
+Note that the size is determined by the size of Ptr -- if we are in memory tracking mode, then a Ptr will contain more than just a ptr.  
+
+Unlike BitVector, there isn't one piece of extremely important code.  Instead, there are two, those being the functions Set and SetRef.
+
+```c++
+/// Set a variable by ID.
+template <typename  T> T  &  Set(size_t  id, const  T  &  value) {
+	if (memory.isRef(id)) {
+		memory.Destruct<T>(id);
+		memory.setRef(id, false);
+		new (memory.GetPtr<T>(id+1).Raw()) T(value);
+		return  memory.Get<T>(id);
+	} else {
+		return (Get<T>(id) = value);
+	}
+}
+
+/// Set a reference to a variable by ID.
+template <typename  T> T  &  SetRef(size_t  id, T  &  value) {
+	if (!memory.isRef(id)) {
+		memory.Destruct<T>(id);
+		memory.setRef(id, true);
+		new (memory.GetPtr<Ptr<T>>(id+1).Raw()) Ptr<T>(& value);
+	} else {
+		*memory.GetPtr<Ptr<T>>(id+1) = & value;
+	}
+	return  memory.Get<T>(id);
+}
+```
+
+The complex bit of this code is the "new" phrase.  It creates an instance of the object using the copy constructor at where the pointer to the object (which is actually a reference, but that doesn't matter since it's been destructed).  Turns out you need to do it this way or you end up with memory leaks.  That was probably the most difficult part of this project.
```

Comment:
Before getting into the complex bit of code I would maybe include a sentence or too as to how the other bits of code is working (maybe as comments in the actual code chunk itself)